File: java/BestMatching.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class BestMatching {

    private static Map<String, Integer> getWordFrequency(String text) {
        Map<String, Integer> frequencyMap = new HashMap<>();
        for (String word : text.toLowerCase().split("\s+")) {
            frequencyMap.put(word, frequencyMap.getOrDefault(word, 0) + 1);
        }
        return frequencyMap;
    }

    private static double cosineSimilarity(Map<String, Integer> freq1, Map<String, Integer> freq2) {
        Set<String> allWords = new HashSet<>(freq1.keySet());
        allWords.addAll(freq2.keySet());

        int dotProduct = 0;
        double norm1 = 0;
        double norm2 = 0;

        for (String word : allWords) {
            int count1 = freq1.getOrDefault(word, 0);
            int count2 = freq2.getOrDefault(word, 0);

            dotProduct += count1 * count2;
            norm1 += Math.pow(count1, 2);
            norm2 += Math.pow(count2, 2);
        }

        return dotProduct / (Math.sqrt(norm1) * Math.sqrt(norm2));
    }

    public static Record findBestMatchingParagraph(List<Record> records, String inputSentence) {
        Record bestMatch = null;
        double highestSimilarity = 0;
        Map<String, Integer> inputFreqMap = getWordFrequency(inputSentence);

        for (Record record : records) {
            Map<String, Integer> paragraphFreqMap = getWordFrequency(record.paragraphText);
            double similarity = cosineSimilarity(inputFreqMap, paragraphFreqMap);
            record.similarity = similarity;

            if (similarity > highestSimilarity) {
                highestSimilarity = similarity;
                bestMatch = record;
            }
        }

        return bestMatch;
    }
}

File: go/best_matching.go
package main

import (
	"math"
	"strings"
)


// getWordFrequency calculates the word frequency map for a given text
func getWordFrequency(text string) map[string]int {
	frequencyMap := make(map[string]int)
	words := strings.Fields(strings.ToLower(text)) // Split text into words and convert to lowercase
	for _, word := range words {
		frequencyMap[word]++
	}
	return frequencyMap
}

// cosineSimilarity calculates the cosine similarity between two word frequency maps
func cosineSimilarity(freq1, freq2 map[string]int) float64 {
	allWords := make(map[string]bool)
	for word := range freq1 {
		allWords[word] = true
	}
	for word := range freq2 {
		allWords[word] = true
	}

	dotProduct := 0
	norm1, norm2 := 0.0, 0.0

	for word := range allWords {
		count1 := freq1[word]
		count2 := freq2[word]

		dotProduct += count1 * count2
		norm1 += float64(count1 * count1)
		norm2 += float64(count2 * count2)
	}

	if norm1 == 0 || norm2 == 0 {
		return 0 // Avoid division by zero
	}
	return float64(dotProduct) / (math.Sqrt(norm1) * math.Sqrt(norm2))
}

// FindBestMatchingParagraph finds the record with the highest similarity to the input sentence
func FindBestMatchingParagraph(records []Record, inputSentence string) *Record {
	var bestMatch *Record
	highestSimilarity := 0.0

	inputFreqMap := getWordFrequency(inputSentence)

	for i := range records {
		paragraphFreqMap := getWordFrequency(records[i].ParagraphText)
		similarity := cosineSimilarity(inputFreqMap, paragraphFreqMap)
		records[i].Similarity = similarity

		if similarity > highestSimilarity {
			highestSimilarity = similarity
			bestMatch = &records[i]
		}
	}

	return bestMatch
}